runner/src/runner.rs

```diff
@@ -0,0 +1,628 @@
+    }
+
+    // Declaring as a static method would highly complicate mocking
+    async fn get_request_bytes(&self, url: String) -> Result<Bytes, Error> {
+        let response = reqwest::get(url.clone()).await?;
+        Ok(response.bytes().await?)
+    }
+}
+
+#[async_trait]
+pub trait StorageReader {
+    fn compute_storage_key(&self, module: String, key: String) -> String;
+    async fn query_storage(
+        &self,
+        parachain_rpc: &WebsocketClient,
+        maybe_storage_key: Option<String>,
+        method: String,
+    ) -> Option<SpCoreBytes>;
+    async fn read_chain_storage<T: 'static + Decode + Debug>(
+        &self,
+        parachain_rpc: &WebsocketClient,
+        maybe_storage_key: Option<String>,
+    ) -> Result<Option<T>, Error>;
+}
+
+#[async_trait]
+impl StorageReader for Runner {
+    fn compute_storage_key(&self, module: String, key: String) -> String {
+        let module = twox_128(module.as_bytes());
+        let item = twox_128(key.as_bytes());
+        let key = hex::encode([module, item].concat());
+        format!("0x{}", key)
+    }
+
+    async fn query_storage(
+        &self,
+        parachain_rpc: &WebsocketClient,
+        maybe_storage_key: Option<String>,
+        method: String,
+    ) -> Option<SpCoreBytes> {
+        let params = maybe_storage_key.map_or(rpc_params![], |key| rpc_params![key]);
+        parachain_rpc.inner().request(method.as_str(), params).await.ok()
+    }
+
+    async fn read_chain_storage<T: 'static + Decode + Debug>(
+        &self,
+        parachain_rpc: &WebsocketClient,
+        maybe_storage_key: Option<String>,
+    ) -> Result<Option<T>, Error> {
+        read_chain_storage(self, parachain_rpc, maybe_storage_key).await
+    }
+}
+
+pub async fn ws_client(url: &str) -> Result<WebsocketClient, Error> {
+    let ws_client = WsClientBuilder::default().build(url).await?;
+    Ok(WebsocketClient::new(ws_client))
+}
+
+async fn download_binary(runner: &impl RunnerExt, release: ClientRelease) -> Result<DownloadedRelease, Error> {
+    let (bin_name, bin_path) = runner.uri_to_bin_path(&release.uri)?;
+    log::info!("Downloading {} at: {:?}", bin_name, bin_path);
+    File::create(bin_path.clone())?;
+
+    let bytes = runner.get_request_bytes(release.uri.clone()).await?;
+    fs::write(&bin_path, &bytes)?;
+
+    // Make the binary executable.
+    // The set permissions are: -rwx------
+    fs::set_permissions(bin_path.clone(), fs::Permissions::from_mode(0o700))?;
+
+    Ok(DownloadedRelease {
+        release,
+        path: bin_path,
+        bin_name: bin_name.to_string(),
+    })
+}
+
+fn uri_to_bin_path(runner: &impl RunnerExt, uri: &str) -> Result<(String, PathBuf), Error> {
+    // Remove any trailing slashes from the release URI
+    let parsed_uri = Url::parse(uri.trim_end_matches('/'))?;
+    let bin_name = parsed_uri
+        .path_segments()
+        .and_then(|segments| segments.last())
+        .and_then(|name| if name.is_empty() { None } else { Some(name) })
```

Comment:
What does `segments.last()` return? Is it already an option?

Reply:
As I understand it, `segments.last()` gives a `Option<string>`, and then this line filters empty strings, i.e. for when `segments.last` returns `Some("")`